components/authz-service/server/v2/project.go

```diff
@@ -398,17 +397,49 @@ func fromStorageConditions(cs []storage.Condition) ([]*api.Condition, error) {
 }
 
 func fromStorageCondition(c storage.Condition) (*api.Condition, error) {
-	// TODO: stringly-typed. meh.
-	t, ok := api.ProjectRuleConditionTypes_value[strings.ToUpper(c.Type.String())]
-	if !ok {
-		return nil, fmt.Errorf("invalid condition type %s", c.Type.String())
+	t, err := fromStorageConditionType(c.Attribute)
+	if err != nil {
+		return nil, err
 	}
 	return &api.Condition{
-		Type:   api.ProjectRuleConditionTypes(t),
+		Type:   t,
 		Values: c.Value,
 	}, nil
 }
 
+var storageToAPIConditionAttributes = map[storage.ConditionAttribute]api.ProjectRuleConditionTypes{
+	storage.ChefRole:     api.ProjectRuleConditionTypes_ROLES,
+	storage.ChefServer:   api.ProjectRuleConditionTypes_CHEF_SERVERS,
+	storage.ChefTag:      api.ProjectRuleConditionTypes_CHEF_TAGS,
+	storage.Environment:  api.ProjectRuleConditionTypes_CHEF_ENVIRONMENTS,
+	storage.Organization: api.ProjectRuleConditionTypes_CHEF_ORGS,
+	storage.PolicyGroup:  api.ProjectRuleConditionTypes_POLICY_GROUP,
+	storage.PolicyName:   api.ProjectRuleConditionTypes_POLICY_NAME,
+}
+
+var apiToStorageConditionAttributes map[api.ProjectRuleConditionTypes]storage.ConditionAttribute
+var onceReverseConditionAttributesMapping sync.Once
```

Comment:
what does `sync.Once` do?

Reply:
Exactly! It's the lazy way for not having to maintain both mappings manually. Only A -> B is hardcoded, B -> A is initialized _once_ in the way @tylercloke described.